### PR TITLE
Log rejected causes when trying to mark promises as failed

### LINKED ISSUE #631: Better log errors that should be reported by `tryFailure` but aren't
Explicitly I seem to be on a network where the connection-close timeout beats the dns resolution timeout but both happen.

This causes the generic `java.lang.IllegalStateException: Channel closed before HTTP/2 preface completed.` to beat the much more specific `java.net.UnknownHostException: failed to resolve 'api.development.push.apple.com' after 4 queries` in `com.turo.pushy.apns.ApnsChannelFactory#create`. I had to debug the code with a breakpoint there to find the cause of the issue.

Would it be possible to change anywhere that `promise.tryFailure(ex)` is called it can be changed to
```java
if (!promise.tryFailure(ex)) {
  logSwallowedError(ex);
}
```

### PR DESCRIPTION
**Pull request by jchambers**
This closes #631 by logging promise failure causes in cases where we try to mark the promise as failed, but the promise was already done for some other reason.

@mattnathan is this what you had in mind?

### CODE REVIEW

File: pushy/src/main/java/com/turo/pushy/apns/ApnsChannelFactory.java [hunk 1]
```diff
@@ -249,4 +253,10 @@ public void close() {
             }
         }
     }
+
+    private static void tryFailureAndLogRejectedCause(final Promise<?> promise, final Throwable cause) {
+        if (!promise.tryFailure(cause)) {
+            log.debug("Could not set failure cause for promise because promise was already done. Rejected cause was {}", cause, cause);
```
Comment (jchambers) [inline pushy/src/main/java/com/turo/pushy/apns/ApnsChannelFactory.java]: This logs both a summary of the exception AND a stack trace, which might be overkill. Very open to suggestions here.

**Comment by jchambers** [comment]
Also worth pointing out: I narrowly scoped this to `ApnsChannelFactory`, though it's arguably useful in `ApnsChannelPool`, `ApnsClientHandler`, and `ApnsClient`, too. Having reviewed those cases, I think we'll get the highest benefit-to-noise ratio by focusing on `ApnsChannelFactory`, but am happy to be convinced otherwise!

**Comment by jchambers** [comment]
…I also delayed when we try the vague "connection closed" failure reason in 788c841. That doesn't change any of the fundamentals of the issue, but should at least increase the probability that we'll have a more specific error message to offer.

**Comment by jchambers** [comment]
Hm. Now I'm actually thinking that 788c841 is just asking for trouble and am inclined to revert it.

**Comment by jchambers** [comment]
I had another idea, but I don't think it'll work. We _could_ use [`Throwable#addSuppressed(Throwable)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Throwable.html#addSuppressed(java.lang.Throwable)) to add the "bounced" exception as a suppressed exception to whatever we get from `Promise#cause()`. The trouble is that it's _still_ a race condition. Callers might react to the failed `Promise` before we add the suppressed exception, and would never see that anything was added after the fact.

Sadly, it seems like logging is still the best option we have.

**Comment by mattnathan** [comment]
Sorry, have been on annual leave and haven't checked email/github. The change looks good. 

I might say that a Warning rather than Debug log would be better but I leave that decision up to you :)

**Comment by jchambers** [comment]
No worries! Thanks for getting back to me.

I did go back and forth on the logging level, but in keeping with the [logging policy in the README](https://github.com/relayrides/pushy#logging), I didn't feel like this rose to the level of a "serious error." I could definitely be persuaded, though!

**Comment by mattnathan** [comment]
In my case it would fit into the "errors that may indicate a bug in caller's code" - though more bug in the callers system, i.e. bad dns setup. 

I guess the reason it's not sitting right with me is that without knowledge of how the pushy code works you wouldn't know that a 'closed early' type error could have more information in the debug log relating to a suppressed exception, causing undue headaches for the dev until they finally get around to enabling debug logging, probably after checking the source code like I did.

Checking Guava (because I tend to think it's a well designed lib) I notice that anywhere exceptions are suppressed it's logged as a warning.

**Comment by jchambers** [comment]
Yeah—that makes sense to me. I'll bump it up to `warn`.

**Pull request merged**